gh-135836: Fix IndexError in asyncio.create_connection with empty exceptions list

heliang666s opened the pull request:
Fix IndexError that occurs in asyncio.create_connection() when Happy Eyeballs algorithm leaves empty sublists in the exceptions list, which after flattening becomes an empty list causing str(exceptions[0]) to crash. The fix adds explicit handling for empty exceptions list by changing 'else:' to 'elif len(exceptions) > 1:' and adding a new else clause that raises OSError('create_connection failed') instead of crashing.


<!-- gh-issue-number: gh-135836 -->
* Issue: gh-135836
<!-- /gh-issue-number -->

heliang666s committed:
Fix IndexError in asyncio.create_connection with empty exceptions list

Fix IndexError that occurs in asyncio.create_connection() when Happy Eyeballs algorithm leaves empty sublists in the exceptions list, which after flattening becomes an empty list causing str(exceptions[0]) to crash. The fix adds explicit handling for empty exceptions list by changing 'else:' to 'elif len(exceptions) > 1:' and adding a new else clause that raises OSError('create_connection failed') instead of crashing.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

heliang666s committed:
Fix NEWS entry for gh-135836 - correct function reference

Changed :func: to double backticks to fix Sphinx reference error

heliang666s committed:
Re-trigger CI

serhiy-storchaka commented on Lib/asyncio/base_events.py:
```diff
@@ -1170,6 +1170,9 @@ async def create_connection(
                             ', '.join(str(exc) for exc in exceptions)))
+                    else:
+                        # No exceptions were collected, raise a generic connection error
+                        raise OSError('create_connection failed')
```
Would not be better to raise a `TimeoutError`? This only happens when all connections exceed their time.

serhiy-storchaka reviewed (commented):
The approach LGTM, but please add a test. You can take `test_create_connection_timeout` as example. It only tests for `happy_eyeballs_delay=None` (default), we need a test for non-None value.

heliang666s committed:
Address core dev feedback: use TimeoutError instead of OSError

Core developer suggested TimeoutError is more appropriate when empty
exceptions list indicates all connections exceeded their timeout.

ZeroIntensity commented on Misc/NEWS.d/next/Library/2025-06-23-11-04-25.gh-issue-135836.-C-c4v.rst:
```diff
@@ -0,0 +1 @@
+Fixed :exc:`IndexError` in ``asyncio.create_connection()`` that could occur when the Happy Eyeballs algorithm resulted in an empty exceptions list during connection attempts.
```
Active voice and use a Sphinx ref:
```suggestion
Fix :exc:`IndexError` in :meth:`asyncio.loop.create_connection` that could occur when the Happy Eyeballs algorithm resulted in an empty exceptions list during connection attempts.
```

heliang666s committed:
Merge branch 'python:main' into fix-asyncio-create-connection-indexerror

heliang666s committed:
Add test asyncio.create_connection Happy Eyeballs empty exceptions and update NEWS entry.

heliang666s committed:
Test asyncio.create_connection Happy Eyeballs empty exceptions and update NEWS entry.

kumaraditya303 commented on Lib/test/test_asyncio/test_base_events.py:
```diff
@@ -1190,6 +1190,37 @@ def getaddrinfo(*args, **kw):
+                MyProto, 'example.com', 80, happy_eyeballs_delay=0.1)
+
+            # Should raise TimeoutError instead of IndexError
+            with self.assertRaises(TimeoutError):
```
Can you add check for the error message as well?

heliang666s committed:
test: check for error message in test_create_connection_happy_eyeballs_empty_exceptions

heliang666s committed:
Merge branch 'main' into fix-asyncio-create-connection-indexerror

kumaraditya303 commented on Lib/test/test_asyncio/test_base_events.py:
```diff
@@ -1190,9 +1190,44 @@ def getaddrinfo(*args, **kw):
+                MyProto, 'example.com', 80, happy_eyeballs_delay=0.1)
+
+            # Should raise TimeoutError instead of IndexError
+            with self.assertRaisesRegex(TimeoutError, "connection timed out"):
```
The msg should be "create_connection failed" no?

heliang666s committed:
test: fix test_create_connection_happy_eyeballs_empty_exceptions to be synchronous and check correct error message

kumaraditya303 commented on Lib/test/test_asyncio/test_base_events.py:
```diff
@@ -1190,9 +1190,44 @@ def getaddrinfo(*args, **kw):
+        # host, port and sock are specified
         coro = self.loop.create_connection(
-            MyProto, 'example.com', 80, sock=object())
+            MyProto, 'example.com', 80, sock=mock.Mock())
```
Why this change is needed?

kumaraditya303 commented on Lib/asyncio/base_events.py:
```diff
@@ -1161,7 +1161,7 @@ async def create_connection(
                     if len(exceptions) == 1:
                         raise exceptions[0]
-                    else:
+                    elif len(exceptions) > 1:
```
```suggestion
                    elif exceptions:
```

heliang666s committed:
Revert unintended change

heliang666s committed:
Update Lib/asyncio/base_events.py

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

kumaraditya303 commented on Lib/test/test_asyncio/test_base_events.py:
```diff
@@ -1190,6 +1190,36 @@ def getaddrinfo(*args, **kw):
+    @patch_socket
+    def test_create_connection_happy_eyeballs_empty_exceptions(self, m_socket):
+        # Test for gh-135836: Fix IndexError when Happy Eyeballs algorithm
+        # results in empty exceptions list
```
```suggestion
        # See gh-135836: Fix IndexError when Happy Eyeballs algorithm
        # results in empty exceptions list
```

heliang666s committed:
Update Lib/test/test_asyncio/test_base_events.py

Co-authored-by: Kumar Aditya <kumaraditya@python.org>
